db/view/view_update_generator.cc

```diff
@@ -24,6 +25,76 @@ static inline void inject_failure(std::string_view operation) {
 
 namespace db::view {
 
+class view_update_generator::progress_tracker final : public sstables::read_monitor_generator {
+    class read_monitor final : public  sstables::read_monitor {
+        sstables::shared_sstable _sst;
+        const sstables::reader_position_tracker* _tracker = nullptr;
+        uint64_t _last_position_seen = 0;
+    public:
+        virtual void on_read_started(const sstables::reader_position_tracker& tracker) override {
+            _tracker = &tracker;
+        }
+
+        virtual void on_read_completed() override {
+            if (auto tracker = std::exchange(_tracker, nullptr)) {
+                _last_position_seen = tracker->position;
+            }
+        }
+
+        uint64_t pending_work() const noexcept {
+            auto last_pos = (_tracker) ? _tracker->position : _last_position_seen;
+            return _sst->data_size() - last_pos;
+        }
+
+        read_monitor& operator=(const read_monitor&) = delete;
+        read_monitor(const read_monitor&) = delete;
+        read_monitor& operator=(const read_monitor&&) = delete;
+        read_monitor(read_monitor&&) = delete;
+
+        explicit read_monitor(sstables::shared_sstable sst)
+            : _sst(std::move(sst)) {
+        }
+    };
+private:
+    // Tracks SSTables that were registered in view_update_generator, but aren't being processed yet.
+    uint64_t _inactive_pending_work = 0;
+    // Tracks SSTables that are now being processed by view_update_generator's async loop
+    // using unordered_map to provide a stable address for read_monitor, so operator() can safely return a reference.
+    std::unordered_map<sstables::shared_sstable, read_monitor> _monitors;
+public:
+    virtual sstables::read_monitor& operator()(sstables::shared_sstable sst) override {
+        auto p = _monitors.try_emplace(sst, sst);
+        _inactive_pending_work -= sst->data_size();
+        return p.first->second;
+    }
+
+    void on_sstable_registration(const sstables::shared_sstable& sst) {
+        _inactive_pending_work += sst->data_size();
+    }
+
+    void on_sstables_deregistration(const std::vector<sstables::shared_sstable>& ssts) {
```

Comment:
s/deregistration/unregistration/

Reply:
> s/deregistration/unregistration/

if there's nothing else, let's please postpone this rename.

Reply:
Ok, queued